### PR TITLE
frontend: type socket

### PR DESCRIPTION
**Pull request by thisconnect**
Builds on https://github.com/digitalbitbox/bitbox-wallet-app/pull/1769

- typed TMsgCallback for Android, Qt and WebSocket transports
- added events and legacyevents types to type the payload that is
  pushed by the backend
- updated types where raw apiSubscribe is used

### CODE REVIEW

File: frontends/web/src/api/socket.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ */
+type TAction = 'replace' | 'reload';
+
+/**
+ * This type models the events that are received from the backend.
+ */
+export type TEvent = {
+  readonly action?: TAction;
+  readonly object?: any;
+  readonly subject?: TSubject;
```
Comment (benma) [inline frontends/web/src/api/socket.ts:33]: why `?`? It should always be defined

File: frontends/web/src/api/subscribe.ts [hunk 2]
```diff
@@ -29,16 +30,18 @@ export function subscribeEndpoint<T>(
...
     switch (event.action) {
     case 'replace':
       cb(event.object);
       break;
     case 'reload':
-      // TODO: backend should push data with "replace" and not use "reload"
-      apiGet(event.subject)
-        .then(object => cb(object))
-        .catch(console.error);
+      if (event.subject) {
```
Comment (benma) [inline frontends/web/src/api/subscribe.ts:39]: if clause not needed

File: frontends/web/src/decorators/subscribe.tsx [hunk 3]
```diff
@@ -64,13 +64,15 @@ export function subscribe<LoadedProps extends ObjectButNotFunction, ProvidedProp
...
         this.unsubscribeEndpoint(key);
-        this.subscriptions[key] = apiSubscribe(endpoint, (event: Event) => {
+        this.subscriptions[key] = apiSubscribe(endpoint, (event) => {
           switch (event.action) {
           case 'replace':
             this.setState({ [key]: event.object } as Pick<LoadedProps, keyof LoadedProps>);
             break;
           case 'reload':
-            apiGet(event.subject).then(object => this.setState({ [key]: object } as Pick<LoadedProps, keyof LoadedProps>));
+            if (event.subject) {
```
Comment (benma) [inline frontends/web/src/decorators/subscribe.tsx:73]: same, if clause not needed

File: frontends/web/src/api/socket.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+/**
+ * This type models the legacy events that are received from the backend.
+ */
+export type TEventLegacy = {
+  readonly code?: string;
+  readonly data?: TSubject;
+  readonly deviceID?: string;
+  readonly meta?: any;
+  readonly type?: string;
```
Comment (benma) [inline frontends/web/src/api/socket.ts:44]: type is always defined

File: frontends/web/src/api/socket.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ * This type enumerates the various actions of an event.
+ */
+type TAction = 'replace' | 'reload';
+
+/**
+ * This type models the events that are received from the backend.
+ */
+export type TEvent = {
+  readonly action?: TAction;
+  readonly object?: any;
```
Comment (benma) [inline frontends/web/src/api/socket.ts:32]: object is always defined, it can be `null`. but since it it typed as `any` you can just remove the `?`

File: frontends/web/src/api/socket.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * This type enumerates the various actions of an event.
+ */
+type TAction = 'replace' | 'reload';
+
+/**
+ * This type models the events that are received from the backend.
+ */
+export type TEvent = {
+  readonly action?: TAction;
```
Comment (benma) [inline frontends/web/src/api/socket.ts:31]: action is always defined.

File: frontends/web/src/utils/event-legacy.ts [hunk 7]
```diff
@@ -15,61 +15,40 @@
...
  */
-
 const subscriptions: ISubscriptions = {};
 
 /**
  * This function dispatches the events from the websocket to the observers.
  */
-
-function handleMessages(payload: IEvent): void {
+function handleMessages(payload: TPayload): void {
```
Comment (benma) [inline frontends/web/src/utils/event-legacy.ts:45]: not `TEventLegacy`?

File: frontends/web/src/utils/event.ts [hunk 8]
```diff
@@ -60,11 +42,14 @@ const subscriptions: Subscriptions = {};
 /**
  * This function dispatches the events from the websocket to the observers.
  */
-function handleEvent(payload: any): void {
-  if (typeof payload.subject === 'string') {
+function handleEvent(payload: TPayload): void {
```
Comment (benma) [inline frontends/web/src/utils/event.ts:45]: not `TEvent`?

File: frontends/web/src/utils/transport-common.ts [hunk 9]
```diff
@@ -14,11 +14,13 @@
  * limitations under the License.
  */
 
+import { TPayload } from '../api/socket';
```
Comment (benma) [inline frontends/web/src/utils/transport-common.ts:17]: something is wrong here - api types are leaking into the generic transport-common.ts. 

imo in general, the socket files and functions are not very spread out and hard to follow. 

would it make sense to first change every use of `apiWebsocket` to `event-legacy.ts:subscribe` or `event.ts:apiSubscribe`? then the types also don't have to be combined into socket.ts and all the switch cases are reduced/removed. wdyt?

File: frontends/web/src/utils/event-legacy.ts (hunk 7 above)
Comment (thisconnect) [reply frontends/web/src/utils/event-legacy.ts:45]: handleMessages receives both `type TPayload = TEventLegacy | TEvent;`

and then later knows if it is `TEvent` or `TEventLegacy`. I'll make another comment further down where `observer(payload);` is called.

File: frontends/web/src/utils/event.ts (hunk 8 above)
Comment (thisconnect) [reply frontends/web/src/utils/event.ts:45]: nope, same as other comment, after the checks TypeScript knows it is TEvent.

Screen Shot 2022-10-19 at 15 05 54

**Comment by thisconnect** [comment]
@benma do you mind if I rename all `utils/*transport.ts` files to `utils/transport-*.ts` so they exist next to each other?

androidtransport.ts -> transport-android.ts<br>
qttransport.ts -> transport-qt.ts<br>
websocket.ts -> transport-websocket.ts

**Comment by benma** [comment]
> @benma do you mind if I rename all `utils/*transport.ts` files to `utils/transport-*.ts` so they exist next to each other?
> 
> androidtransport.ts -> transport-android.ts qttransport.ts -> transport-qt.ts websocket.ts -> transport-websocket.ts

no feel free!

**Comment by thisconnect** [comment]
I didn't rename `websocket.ts -> transport-websocket.ts` because websocket.ts  has more logic (checks if running in qt or android and uses a different transport with fallback to WebSocket). I think this logic should not be in `transport-websocket.ts` but could be in just `socket.ts`.

**Comment by thisconnect** [comment]
@benma 

> > @benma do you mind if I rename all `utils/*transport.ts` files to `utils/transport-*.ts` so they exist next to each other?
> > androidtransport.ts -> transport-android.ts qttransport.ts -> transport-qt.ts websocket.ts -> transport-websocket.ts
> 
> no feel free!

ok added 1 more commit that moved and renamed types PTAL

**Review by benma — APPROVED** [review]
LGTM. see nit

File: frontends/web/src/utils/transport-websocket.ts [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+import { apiPort, apiToken, isTLS } from './request';
+import { TMsgCallback, TUnsubscribeCallback } from './transport-common';
+
+let socket: WebSocket | undefined;
+
+const currentListeners: TMsgCallback[] = [];
+
+export function webSubscribePushNotifications(msgCallback: TMsgCallback): TUnsubscribeCallback {
```
[the remark this answers is not in the document]
Comment (thisconnect) [reply frontends/web/src/utils/transport-websocket.ts]: fixed
